### PR TITLE
Terraform cleanup

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

It's time for a bit of cleanup. Now that 1.0.0 is out, let's make it the default.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by timstott — APPROVED** [review]
LGTM :tada:

**Comment by zimbatm** [comment]
Result of `nixpkgs-review pr 126919` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>terraform</li>
    <li>terraform-full</li>
    <li>terraform_plugins_test</li>
  </ul>
</details>

**Review by babariviere — APPROVED** [review]
All good

File: pkgs/applications/networking/cluster/terraform/default.nix [hunk 1]
```diff
@@ -127,76 +143,47 @@ let
...
-  terraform_0_15 = pluggable (generic {
-    version = "0.15.5";
-    sha256 = "18f4a6l24s3cym7gk40agxikd90i56q84wziskw1spy9rgv2yx6d";
-    vendorSha256 = "12hrpxay6k3kz89ihyhl91c4lw4wp821ppa245w9977fq09fhnx0";
-    patches = [ ./provider-path-0_15.patch ];
-    passthru = { inherit plugins; };
-  });
-
-  terraform_1_0_0 = pluggable (generic {
+  terraform_1_0_0 = mkTerraform {
```
Comment (flokli) [inline pkgs/applications/networking/cluster/terraform/default.nix:159]: The harm probably is already done, but `terraform_1_0_x` or `terraform_1_0` sounds more future-proof.

We want to be able to bump to a point release, if they do one.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -31242,15 +31242,12 @@ in
   termpdfpy = python3Packages.callPackage ../applications/misc/termpdf.py {};
 
   inherit (callPackage ../applications/networking/cluster/terraform { })
-    terraform_0_12
-    terraform_0_13
-    terraform_0_14
-    terraform_0_15
```
Comment (flokli) [inline pkgs/top-level/all-packages.nix:31248]: We just entirely drop them? In that case, we need to add a `throw` to the aliases, and add a changelog entry.

**Review by flokli — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/cluster/terraform/default.nix (hunk 1 above)
Comment (zimbatm) [reply pkgs/applications/networking/cluster/terraform/default.nix]: good idea. This one slipped previously.

**Comment by endocrimes** [comment]
This feels a bit premature - Especially as not all organizations have been able to move to 0.15 yet - I think we probably should give this a couple of months rather than dropping this - especially as keeping a terraform version in an overlay is _somewhat_ annoying.

(I know I'm not the only person who now wouldn't be able to update their primary system for a little while due to this change, which is quite unfortunate).

**Comment by grahamc** [comment]
re https://github.com/NixOS/nixpkgs/pull/127203

**Comment by grahamc** [comment]
One thing @flokli mentioned but was not addressed:

>  We just entirely drop them? In that case, we need to add a throw to the aliases, and add a changelog entry.

I think this is particularly important since the upgrade from 0.12 to 1.0 is not automatic, and requires non-trivial steps.

**Comment by kalbasit** [comment]
> This feels a bit premature - Especially as not all organizations have been able to move to 0.15 yet - I think we probably should give this a couple of months rather than dropping this - especially as keeping a terraform version in an overlay is _somewhat_ annoying.
> 
> (I know I'm not the only person who now wouldn't be able to update their primary system for a little while due to this change, which is quite unfortunate).

I agree. Can we change the default but leave older versions until next release?

**Comment by zimbatm** [comment]
I might have been a bit too enthusiastic about the cleaning. Luckily git makes it easy to revert things, which is what I propose we do: https://github.com/NixOS/nixpkgs/pull/127239

**Comment by zowoq** [comment]
One year on, shall we try this again?

**Comment by zimbatm** [comment]
Sounds good!

**Pull request merged**